Lightweight Sinc Convolutions for Espnet2

lumaku opened the pull request:
This pull request implements Lightweight Sinc convolutions for ESPnet 2.
https://arxiv.org/abs/2010.07597
See #2595 for more details and the implementation for ESPnet 1.

You can use sinc convolutions in your model, simply by adding the following lines to your yaml configuration file:

```yaml
frontend: sliding_window
frontend_conf:
    hop_length: 200
preencoder: sinc
```

Remarks:
As the input format is different to the default frontend, the step of calculating the global norm has to be re-done when switching from the default frontend, which is currently stage 9 in `asr.sh`. The default `hop_length` is set to the kaldi default, but depending on the task and model, ASR performance or memory consumption can be improved by finetuning this setting. This code was tested on RNN-based models.

#2595

kamo-naoyuki commented:
Could add test codes?

kamo-naoyuki commented on espnet2/asr/frontend/sinc.py:
```diff
@@ -0,0 +1,229 @@
+#!/usr/bin/env python3
+# encoding: utf-8
```
`encoding: utf-8` is meaning less in python3, and it's not recommended by PEP

lumaku replied:
I removed those lines.

kamo-naoyuki commented on espnet2/asr/frontend/sinc.py:
```diff
@@ -0,0 +1,229 @@
+    ):
+        """Generate a block for lightweight Sinc convolutions.
+
+        :param in_channels:  Number of input channels
```
We are using google docstring style, please change to it.

https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html

lumaku replied:
Changed!

kamo-naoyuki commented on espnet2/asr/frontend/sinc.py:
```diff
@@ -0,0 +1,229 @@
+    def __init__(
+        self,
+        fs: Union[int, str] = 16000,
+        in_channels=1,
```
please add type annotation for all functions.

lumaku replied:
OK

kamo-naoyuki commented on espnet2/tasks/asr.py:
```diff
@@ -53,7 +55,7 @@
 frontend_choices = ClassChoices(
     name="frontend",
-    classes=dict(default=DefaultFrontend),
+    classes=dict(default=DefaultFrontend, raw=SlidingWindow),
```
Why `raw`? 
Should be `sliding_window`

lumaku replied:
Changed to `sliding_window`.

sw005320 commented on espnet2/layers/sinc_conv.py:
```diff
@@ -0,0 +1,181 @@
+import torch
+from typing import Union
+
+CONSTANT_PI = 3.141592653589793
```
I'm curious why you represent \pi as follows.
I just thought that you can use `math.pi`.

lumaku replied:
Correct!
In a previous version, we used `np.pi` and there is not yet a `torch.pi`.
My thought here was that setting the constant is faster than importing the numpy library only for that one number....
Also, one dependency less.

kamo-naoyuki replied:
Please use math.pi.
Not faster, it's always same.

lumaku replied:
OK

sw005320 reviewed (approved):
LGTM!
This is a very good example of learnable feature extraction!!!

Can you put the current result in https://github.com/espnet/espnet/blob/master/egs2/voxforge/asr1/README.md ?
I'm curious.

Also, are you planning to apply it to the other recipes as you did in your paper?
We could do it and investigate more in later PRs including the use of transformer/conformer ASR backend.

kamo-naoyuki commented on espnet2/utils/plot_sinc_filters.py:
```diff
@@ -0,0 +1,260 @@
+#!/usr/bin/env python3
+#  2020, Technische Universität München; Nicolas Lindae, Ludwig Kürzinger
+#  Apache 2.0  (http://www.apache.org/licenses/LICENSE-2.0)
+
```
The path is not correct.
Move this script to pyscripts.

lumaku replied:
moved `espnet2/utils/plot_sinc_filters.py`
to `egs2/TEMPLATE/asr1/pyscripts/utils/plot_sinc_filters.py`.
Is that the correct path?

kamo-naoyuki replied:
Yes, please.

kamo-naoyuki commented on espnet2/utils/plot_sinc_filters.py:
```diff
@@ -0,0 +1,260 @@
+    plot_filters(np.argsort(F_maxs - F_mins), "filters_len_sort.%s" % args.filetype)
+    plot_filters(np.argsort(F_mins), "filters_min_sort.%s" % args.filetype)
+
+    img_path = os.path.join(args.out_folder, "filtergraph.%s" % args.filetype)
```
Please use pathlib instead of os.path

lumaku replied:
OK

kamo-naoyuki commented on espnet2/utils/plot_sinc_filters.py:
```diff
@@ -0,0 +1,260 @@
+    plot_filter_kernels(filters, sample_rate, args)
+
+    def plot_filters(indices, filename):
+        x = np.linspace(0, np.max(F_maxs), np.max(F_maxs) + 1)
```
Please do not use internal function. 

Basically,  internal function shouldn't be used if it's not necessary

lumaku replied:
I refactored the plotting functions.
There are then two nested functions `mel()` and `hz()` left. They depend on how parameter initialization is implemented. It is just easier to keep them nested instead of programming different versions.

kamo-naoyuki commented on espnet2/utils/plot_sinc_filters.py:
```diff
@@ -0,0 +1,260 @@
+    model = torch.load(model_path, map_location="cpu")
+    if "model" in model:  # snapshots vs. model.acc.best
+        model = model["model"]
+    filters = model["preencoder.filters.f"].detach().cpu().numpy()
```
Add type check for preencoder.

lumaku replied:
Added directly a type check for the filter values for more flexibility.

kamo-naoyuki replied:
I can't allow such flexibility. It's d dangerous design.
Please add type check. We must add type check if new class is added.

kamo-naoyuki replied:
Oh, sorry, you don't instantiated the model class, so we can't check the type.
Could you add checking the existence of the attribute?

lumaku replied:
Added the check for the key in the model dictionary.

kamo-naoyuki commented on espnet2/layers/sinc_conv.py:
```diff
@@ -0,0 +1,181 @@
+            window_func: Window function on the filter, one of ["hamming", "none"].
+            fs (str, int, float): Sample rate of the input data
+        """
+        super(SincConv, self).__init__()
```
super().__int__()

This is python3 project, please care not to use python2 syntax.

lumaku replied:
Changed.

kamo-naoyuki commented on espnet2/asr/frontend/sinc.py:
```diff
@@ -0,0 +1,242 @@
+            out_channels: Number of output channels (for each input channel).
+            activation_type: Choice of activation function.
+            dropout_type: Choice of dropout function.
+        """
```
please add  `assert check_argument_types()` for all __init__()

lumaku replied:
Added

kamo-naoyuki commented on espnet2/tasks/asr.py:
```diff
@@ -74,6 +76,15 @@
+    classes=dict(
+        sinc=LightweightSincConvs,
+    ),
+    type_check=AbsFrontend,
```
This is not AbsFrontend.
Please define AbsPreEncoder

lumaku replied:
OK

lumaku commented:
I will add recipes and results to the pull request later, after the code and those crazy CI tests are stable, or maybe in a separate pull request.

Fhrozen reviewed (commented):
LGTM, @lumaku Thank you for your collaboration.
About the test failed, it is because the code is consuming so much time (more than 2 secs).
I will recommend reducing the number of parameters, length of the sequence, or reduce the sampling frequency to make this faster.
Maybe using large stride will help.

kamo-naoyuki commented on test/espnet2/utils/test_plot_sinc_filters.py:
```diff
@@ -0,0 +1,44 @@
+import torch
+
+import egs2.TEMPLATE.asr1.pyscripts.utils.plot_sinc_filters as psf
+from espnet2.layers.sinc_conv import SincConv
```
We don't have importing system for pyscripts, so could you change this file to a script in `test_utils`?

kamo-naoyuki commented:
> I refactored the plotting functions.
> There are then two nested functions mel() and hz() left. They depend on how parameter initialization is implemented. It is ?just easier to keep them nested instead of programming different versions.

I can't understand. What is "how parameter initialization is implemented"?

kamo-naoyuki commented on egs2/TEMPLATE/asr1/pyscripts/utils/plot_sinc_filters.py:
```diff
@@ -0,0 +1,318 @@
+    sample_rate = args.sample_rate
+
+    model = torch.load(model_path, map_location="cpu")
+    if "model" in model:  # snapshots vs. model.acc.best
```
```suggestion
    if isinstance(model, dict) and "model" in model:  # snapshots vs. model.acc.best
```

Now this script doesn't work if it's state.

kamo-naoyuki replied:
Sorry, I missed, please ignore.

kamo-naoyuki commented on espnet2/asr/preencoder/sinc.py:
```diff
@@ -0,0 +1,245 @@
+        self._create_sinc_convs()
+        self.init_sinc_convs()
+        # Sinc filters require custom initialization
+        self.espnet_initialization_fn = self.init_sinc_convs
```
This is a bad way. It's enexpected and there are no reason to set method dynamically.

Please change `espnet_initialization_fn` to a defined method and remove init_sinc_convs.

```
def espnet_initialization_fn
```

lumaku commented on espnet2/asr/preencoder/sinc.py:
```diff
@@ -0,0 +1,245 @@
+        return output_frames, input_lengths  # no state in this layer
+
+    def get_odim(self, idim=400):
+        """Get output dimension by making one inference.
```
Very useful for debugging, however, not necessary for training/inference. I will move this function to the testing script.

lumaku commented:
@Fhrozen Thanks, that explains why the tests have a timeout. As the timeout happened during a matplotlib plot, I assumed that it had something to do with the plotting backend.......

@kamo-naoyuki Instead of using mel scale to initialize, one could use, say, any linear function. then the inverse `hz()` would look different. I'll out-group the mel scale to a separate module instead of using nested functions.

lumaku commented:
@kamo-naoyuki @Fhrozen The code and tests are stable. Please review.

@kamo-naoyuki Regarding the nested functions `mel()` and `hz()`, I changed this to a separate module that maybe fits better as a namespace. See the MelScale and BarkScale modules. The Bark scale has wider filters at start, from the observation on the sinc filters is that they tend to get wider in training (see paper).

Also, I added results on voxforge - a relatively small corpus, and I observed that the net overfitted a little bit, but good enough for testing. I'll maybe add more recipes and results in a later pull request. Any recipe can be switched to Sinc convolutions by adding the above-described block to the yaml file.

kamo-naoyuki reviewed (approved):
LGTM. Many thanks for reflecting my comments! @lumaku
